lib/leanplum_api/api.rb

```diff
@@ -210,21 +222,32 @@ def development_connection
 
     # Deletes the user_id and device_id key/value pairs from the hash parameter.
     def extract_user_id_or_device_id_hash!(hash)
-      user_id = hash.delete(:user_id)
-      device_id = hash.delete(:device_id)
+      user_id = hash.delete(:user_id) || hash.delete(:userIid)
+      device_id = hash.delete(:device_id) || hash.delete(:deviceId)
       fail "No device_id or user_id in hash #{hash}" unless user_id || device_id
 
       user_id ? { userId: user_id } : { deviceId: device_id }
     end
 
+    def fix_iso8601(attr_hash)
+      attr_hash = HashWithIndifferentAccess.new(attr_hash)
+      attr_hash.each { |k, v| attr_hash[k] = v.iso8601 if v.is_a?(Date) || v.is_a?(Time) || v.is_a?(DateTime) }
+      attr_hash
+    end
+    
     # Action can be any command that takes a userAttributes param.  "start" (a session) is the other command that most
     # obviously takes userAttributes.
     # As of 2015-10 Leanplum supports ISO8601 date & time strings as user attributes.
-    def build_user_attributes_hash(user_hash, action = 'setUserAttributes')
-      user_hash = HashWithIndifferentAccess.new(user_hash)
-      user_hash.each { |k, v| user_hash[k] = v.iso8601 if v.is_a?(Date) || v.is_a?(Time) || v.is_a?(DateTime) }
+    def build_user_attributes_hash(user_hash)
+      user_hash = fix_iso8601(user_hash)
+      user_attributes_hash = extract_user_id_or_device_id_hash!(user_hash).merge(action: 'setUserAttributes', userAttributes: user_hash)
+      user_attributes_hash[:devices] = user_hash[:devices] if user_hash[:devices].present?
```

Comment:
why is this necessary?  `user_attributes_hash` is already `user_hash`, just with ISO adjustments, and `user_id` removed, right?

Reply:
oh wait, i see.  shouldn't it be.  i think this isn't quite right though, because you need to remove the `:devices` key from the hash _before_ you create the `user_attributes_hash`.   otherwise you've got the `devices:` in 2 places
```
{
  userAttributes: {
    attribute: 'val',
    devices: devices_hash
  },
  devices: devices_hash
}
```

Reply:
its about the devices hash.
https://www.leanplum.com/docs/api/production#setuserattributes
devices has to be at the top level, not embedded in userAttributes